Change layout of homepage

patrickpatrickpatrick opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Change the layout of homepage so that:

- 'Popular on GOV.UK' links now appear in a 3 column grid layout on desktop
- 'Services and information' and "Featured" sections are now next to each other on desktop.
- The 'Featured' section now uses the new "Two Thirds" variation of the image-card component

## Why

As part of several incremental changes being made to the homepage.

[Trello Card](https://trello.com/c/5ZGhnozF/2024-live-test-5-layout-popular-grid-etc-design-dev-work-live-25-sep-m), [Jira issue NAV-3270](https://gov-uk.atlassian.net/browse/NAV-3270)

## Visual Changes

### Before

![homepage-desktop-before](https://github.com/alphagov/frontend/assets/28779939/081e1312-0752-44f3-a683-4ea2dd69ae1d)

### After

![homepage-after](https://github.com/alphagov/frontend/assets/28779939/cfbaa79e-6a4a-4ba1-99bf-c0ddc24c8ccf)

patrickpatrickpatrick committed:
Change layout of homepage

This is part of a live test of several incremental changes being made to the homepage.

- Services and information and Featured sections are now in a grid layout.
- Font size of the header of the promotion slot cards has been changed to 's', the same size as those in the Services and information section.
- Use the new image-card two thirds variation for promotion slots
- Change Popular on GOV.UK links to a 3 column grid layout

hannako commented on app/assets/config/manifest.js:
```diff
@@ -12,6 +12,7 @@
 //= link views/_cookie-settings.css
 //= link views/_csv_preview.css
 //= link views/_homepage.css
+//= link views/_homepage_new.css
```
I'm finding this second commit a bit hard to follow. I thiiiiink what's happening is you're undoing the css styles you made in the first commit in `homepage.css` , and making those changes in a new file called `homepage_new.css` instead, so that you can make use of the @new_design toggle? But then there are also fresh changes in this commit. So I was wondering if you could rework the first two commits to make it easier to follow. 

BUT I'm also confused as to why we're using the toggle here at all. I was under the impression we'd be using it to incrementally roll out the look and feel changes which is going live mid October - where as this PR is to go live on Monday. Plus you've picked the commits from other cards so it's all rolled into one anyhow. Soooo can we just forget about the toggle in this PR?

Huge apologies if I've totally missed the point of what you're doing here Patrick!

hannako replied:
Ahhh ok so I think you've added this logic (and new homepage css file) because some of the changes form part of the look and feel designs and monica is wanting to include them early, so you're setting the foundations for being able to use the backend toggle later. But I think that will make the roll out of the layout changes for Monday complicated as we'll need to move just these changes out of the the `if` block. 

So if it's ok with you, I think we should remove that logic out of this PR. And we can add it back when/if we need it in subsequent work

MartinJJones replied:
I've removed the logic to set the new styles, so the new homepage layout will be used when the changes are merged

hannalaakso reviewed (commented):
Some minor comments on the CSS. Still need to review the markup changes.

MartinJJones commented on app/assets/stylesheets/views/_homepage.scss:
```diff
@@ -277,19 +270,21 @@
   }
 
-  & > li {
+  & > div li {
```
I've updated the styles in a recent commit to use `.homepage-most-viewed-list__item` class instead - https://github.com/alphagov/frontend/pull/3743/commits/8c4fc5b20ce461db1b393b1a8e4fea3500cc7e8c

MartinJJones commented on app/assets/stylesheets/views/_homepage.scss:
```diff
@@ -228,13 +209,21 @@
   padding: govuk-spacing(6) 0 0;
+
+  @include govuk-media-query($from: "desktop") {
+    padding-right: 32px;
```
I've updated this to use the `govuk-spacing` mixin

MartinJJones commented on app/assets/stylesheets/views/_homepage.scss:
```diff
@@ -245,6 +234,10 @@
 }
 
+.homepage-section__container {
+  padding-right: 32px;
```
This class name and styling was removed in a recent commit as it is not required

hannalaakso commented on app/views/homepage/_links_and_search.html.erb:
```diff
@@ -31,45 +31,52 @@
+        <% t("homepage.index.popular_links").each_with_index do | item, index | %>
+
+          <% if index % 3 == 0 %>
+            <div class="govuk-grid-row">
```
Using a `<div>` here between `<ul>` and `<li>` results in a HTML validation error:
> Element [div](https://html.spec.whatwg.org/multipage/#the-div-element) not allowed as child of element [ul](https://html.spec.whatwg.org/multipage/#the-ul-element) in this context. 

I think we might need to find a different way of building the grid, possibly using CSS grid like we do with the [cards component](https://components.publishing.service.gov.uk/component-guide/cards). Unfortunately that component doesn't currently have a variant that doesn't include the arrows 🤔

MartinJJones committed:
Remove top border from image-card

MartinJJones committed:
Copy CSS grid from gem cards to popular links

The CSS is used for the grid is taken from the `cards` component.

hannalaakso reviewed (approved):
Great, thanks both @MartinJJones and @patrickpatrickpatrick for getting this into such a good place 🙌  We should probably now update the PR description as it's now longer reflective of the work, there have been quite a few design changes since.

I tested the page on the browsers and devices we usually test in. Didn't spot any issues, apart from how on mobile there seems to be a bit too much grey space below the Popular links. I checked with Monica who agreed that we should reduce the amount of space so that's something to look at tomorrow.  

<img width="432" alt="Screenshot 2023-09-21 at 17 40 55" src="https://github.com/alphagov/frontend/assets/5007934/cf5a6a03-8136-4e77-ba5e-61aa62019213">


I'm giving this PR a tick to indicate that I'm happy with everything apart from the above mobile spacing issue so whoever reviews the PR tomorrow can focus on reviewing any small fixes we add tomorrow.

MartinJJones committed:
Fix spacing issue on popular links

- Ensures that the total padding-bottom on desktop underneath Popular links is 50px (30px + 20px from the popular links)
- Removed grid-auto-rows: fraction(1) as it is not required

MartinJJones commented:
@hannalaakso thanks for finding and reporting the spacing issue, I've fixed this issue on a recent commit:

<img width="611" alt="Screenshot 2023-09-22 at 10 19 20" src="https://github.com/alphagov/frontend/assets/28779939/e7796930-f110-494d-b276-e8ac6542061c">

MartinJJones committed:
Improve layout of featured cards on small screens

Temp style overrides to the image card component from the gem to:

- Ensure the space between the image card and the text remains at 15px on mobile, this was previously 30px on mobile.
- Ensure the featured image cards have 15px padding to the left and right when this screen width is =< 319px.

MartinJJones committed:
Ensure the featured heading is styled correctly on mobile

matthillco commented:
I've also completed a secondary review of changes made today and it looks good to go.
